File: create/scripts/qa/refresh_vendor_snapshots.py
#!/usr/bin/env python3
"""Refresh rulesets/Sources/vendor from live yfamilys.com URLs."""
import subprocess
import sys
from pathlib import Path

ROOT = Path(__file__).resolve().parents[2]
VENDOR = ROOT / "rulesets" / "Sources" / "vendor"
UA = (
    "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) "
    "AppleWebKit/537.36 (KHTML, like Gecko) Chrome/120.0.0.0 Safari/537.36"
)
PAIRS = [
    ("https://yfamilys.com/module/adultraplus.sgmodule", "adultraplus.sgmodule"),
    ("https://yfamilys.com/module/bili.module", "bili.module"),
    ("https://yfamilys.com/rule/Kemono.list", "yfamilys_Kemono.list"),
    ("https://yfamilys.com/rule/Cloudflare.list", "yfamilys_Cloudflare.list"),
]


def main() -> int:
    VENDOR.mkdir(parents=True, exist_ok=True)
    failed = 0
    for url, name in PAIRS:
        dest = VENDOR / name
        r = subprocess.run(
            ["curl", "-fL", "-m", "90", "-A", UA, "-o", str(dest), url],
            capture_output=True,
        )
        if r.returncode != 0:
            print(f"FAIL {name} ({url})", file=sys.stderr)
            failed += 1
        else:
            print(f"OK   {name} ({dest.stat().st_size} bytes)")
    return 1 if failed else 0


if __name__ == "__main__":
    sys.exit(main())

File: create/qa/refresh_vendor_snapshots.go
package qa

import (
	"fmt"
	"os"
	"path/filepath"
	"time"

	"github.com/nowaytouse/script_hub/create/hub"
)

func RunRefreshVendorSnapshots() int {
	vendorDir := filepath.Join(hub.ROOT, "rulesets/Sources/vendor")
	hub.EnsureDir(vendorDir)

	pairs := []struct {
		url  string
		name string
	}{
		{"https://yfamilys.com/module/adultraplus.sgmodule", "adultraplus.sgmodule"},
		{"https://yfamilys.com/module/bili.module", "bili.module"},
		{"https://yfamilys.com/rule/Kemono.list", "yfamilys_Kemono.list"},
		{"https://yfamilys.com/rule/Cloudflare.list", "yfamilys_Cloudflare.list"},
	}

	failed := 0
	for _, pair := range pairs {
		dest := filepath.Join(vendorDir, pair.name)
		content := hub.SafeDownload(pair.url, 2, 90)
		if content == "" {
			fmt.Printf("FAIL %s (%s)\n", pair.name, pair.url)
			failed++
		} else {
			hub.SafeWriteFile(dest, content, true)
			info, err := os.Stat(dest)
			size := int64(0)
			if err == nil {
				size = info.Size()
			}
			fmt.Printf("OK   %s (%d bytes)\n", pair.name, size)
		}
		time.Sleep(500 * time.Millisecond) // Add sleep to prevent rate limiting
	}

	if failed > 0 {
		return 1
	}
	return 0
}
